### PR TITLE
[Data] Use function name for progress bars in ray dataset

### LINKED ISSUE #31224: [Datasets] Give dataset progress bars descriptive names
### Description

Instead of names like "Map_Batches":

```
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:13<00:00,  2.21s/it]
```

use names that describe what the stage is doing. For example, if you're using a UDF named `resize_and_pad_images`, we could do:

```
resize_and_pad_images: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:13<00:00,  2.21s/it]
```

### Use case

If your workflow has many preprocessing steps, then there are many progress bars. In this case, it's hard to know what's going on, since all the progress bars say more-or-less the same thing.

```
Read->Map_Batches:   6%|███████▊                                                                                                                           | 12/200 [00:23<04:08,  1.32s/it](scheduler +41s) Tip: use `ray status` to view detailed cluster status. To disable these messages, set RAY_SCHEDULER_EVENTS=0.
(scheduler +41s) Adding 5 node(s) of type worker-node-type-0.
Read->Map_Batches:  46%|████████████████████████████████████████████████████████████▎                                                                      | 92/200 [02:24<02:58,  1.65s/it](raylet) Spilled 2325 MiB, 81 objects, write throughput 79 MiB/s. Set RAY_verbose_spill_logs=0 to disable this message.
Read->Map_Batches:  53%|████████████████████████████████████████████████████████████████████▉                                                             | 106/200 [02:41<03:42,  2.37s/it](raylet) Spilled 10045 MiB, 595 objects, write throughput 221 MiB/s.
Read->Map_Batches:  54%|██████████████████████████████████████████████████████████████████████▏                                                           | 108/200 [02:42<02:24,  1.57s/it](raylet) Spilled 10217 MiB, 604 objects, write throughput 220 MiB/s.
Read->Map_Batches: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [04:33<00:00,  1.43s/it]
Map_Batches:  38%|████████████████████████████████████████████████████▋                                                                                    | 77/200 [00:38<03:08,  1.53s/it](raylet) Spilled 16704 MiB, 1130 objects, write throughput 270 MiB/s.
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:28<00:00,  4.31it/s]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:33<00:00,  6.60it/s]
Sort Sample: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:02<00:00, 120.23it/s]
Shuffle Map:  98%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▌   | 195/200 [00:22<00:00, 10.28it/s](scheduler +7m19s) Resized to 24 CPUs.
Shuffle Map: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:23<00:00,  8.48it/s]
Shuffle Reduce:  91%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████            | 182/200 [00:01<00:14,  1.28it/s](scheduler +7m33s) Adding 1 node(s) of type worker-node-type-0.
(scheduler +7m33s) Resized to 56 CPUs.
Shuffle Reduce:  99%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▋ | 198/200 [00:20<00:01,  1.28it/s](scheduler +7m46s) Resized to 88 CPUs.
Shuffle Reduce: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [01:06<00:00,  3.97s/it]
Sort Sample: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:00<00:00, 355.98it/s]
Shuffle Map: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:02<00:00, 85.12it/s]
Shuffle Reduce: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:03<00:00, 55.77it/s]
Map_Batches:   0%|                                                                                                                                                 (raylet) Spilled 35629 MiB, 1939 objects, write throughput 154 MiB/s.
(scheduler +15m6s) Resized to 104 CPUs.
Map_Batches:   0%|▋                                                                                                                                     | 1/200 [08:Map_Batches: 100%|███████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [08:23<00:00, 351.48s/it]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:02<00:00, 75.73it/s]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:03<00:00, 50.06it/s]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 200/200 [00:02<00:00, 80.15it/s]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 193/193 [00:35<00:00,  5.46it/s]
Map_Batches: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 8/8 [00:13<00:00,  2.21s/it]
```

### PR DESCRIPTION
**Pull request by peytondmurray**
## Why are these changes needed?

This PR changes `map_batches` to display an informative name in the progress bar that is created during calls to `ray.data.dataset.Dataset.map_batches` instead of a generic "map_batches" message, so that a user knows what function is being computed for a given display bar. If `fn` is a bound method of a `Preprocessor` subclass, the progress bar displays the name of the class. Otherwise the `MapBatches(fn.__name__)` is displayed; if the `__name__` attribute is not present on the callable passed to `map_batches`, `type(fn)` is used instead; this happens if the callable is a class instance, for example.

## Related issue number

Closes #31224.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Comment by c21** [comment]
Hi @peytondmurray, thanks for opening up the PR! Could you rebase the PR to latest master, and fix the unit test failures? Thanks.

**Comment by peytondmurray** [comment]
Yep, working on the test failures now. Unfortunately it seems like stdout is being examined in many tests across the code base, so anything that calls `map_batches` needs to be modified. Might be something to consider looking at again in the future if we want to refactor some of the tests.

**Comment by peytondmurray** [comment]
Okay, I think test failures appear to be unrelated.

File: python/ray/air/tests/test_dataset_config.py [hunk 1]
```diff
@@ -283,9 +283,9 @@ def checker(shard, results):
         assert results[0] != results[1], results
         stats = shard.stats()
         assert (
-            "Stage 1 read->randomize_block_order->map_batches: 1/1 blocks executed "
-            in stats
-        ), stats
+            "Stage 1 read->randomize_block_order->"
+            "transform_batch: 1/1 blocks executed "
+        ) in stats, stats
```
Comment (scottjlee) [inline python/ray/air/tests/test_dataset_config.py:288]: nit: should `in stats` be within the parentheses here?

File: python/ray/data/dataset.py [hunk 2]
```diff
@@ -646,7 +646,7 @@ def process_next_batch(batch: DataBatch) -> Iterator[Block]:
 
         plan = self._plan.with_stage(
             OneToOneStage(
-                "map_batches",
+                getattr(fn, "__name__", type(fn)),
```
Comment (scottjlee) [inline python/ray/data/dataset.py:649]: In the case where we fall back to `type(fn)` (I assume for example, in the case of a lambda function), would including `"map_batches"` in the output name be helpful to users?

**Review by scottjlee — APPROVED** [review]

File: python/ray/air/tests/test_dataset_config.py [hunk 3]
```diff
@@ -250,7 +250,7 @@ def checker(shard, results):
         # applying the preprocessor on each epoch.
         assert results[0] == results[1], results
         stats = shard.stats()
-        assert "Stage 1 read->map_batches: 1/1 blocks executed " in stats, stats
+        assert "Stage 1 read->_transform_pandas: 1/1 blocks executed " in stats, stats
```
Comment (ericl) [inline python/ray/air/tests/test_dataset_config.py:253]: Could we preserve the original high level operation name like this? `map_batches(_transform_pandas)`

File: python/ray/data/tests/test_dataset.py [hunk 4]
```diff
@@ -1507,26 +1507,26 @@ def test_dataset_repr(ray_start_regular_shared):
...
-        "+- MapBatches\n"
+        "+- <Lambda>\n"
         "   +- Dataset(num_blocks=10, num_rows=10, schema=<class 'int'>)"
     )
     ds = ds.random_shuffle()
     assert repr(ds) == (
         "RandomShuffle\n"
         "+- Filter\n"
-        "   +- MapBatches\n"
+        "   +- <Lambda>\n"
```
Comment (ericl) [inline python/ray/data/tests/test_dataset.py:1522]:
```suggestion
        "   +- MapBatches(<Lambda>)\n"
```

**Review by ericl — CHANGES_REQUESTED** [review]

File: python/ray/air/tests/test_dataset_config.py (hunk 1 above)
Comment (peytondmurray) [reply python/ray/air/tests/test_dataset_config.py:288]: In this case the parentheses are being used to concatenate the string which doesn't fit on one line. Does

```python
assert (
    "Stage 1 read->randomize_block_order->"
    "transform_batch: 1/1 blocks executed "
in stats), stats
```

do the same thing here? This is a really unusual case where I actually don't know what python does - does it concatenate the string before evaluating `<string> in stats`?

After a bit of testing it seems like you can do it either way. To me it makes sense to group the string together first, like this:

```python
assert (<multiline string>) in stats, stats
```

though if you'd like I can change it.

File: python/ray/data/dataset.py (hunk 2 above)
Comment (peytondmurray) [reply python/ray/data/dataset.py:649]: So the one time I can see where this comes up is not when you pass in a lambda function (`x = lambda y: y; x.__name__` returns `<lambda>`) but rather when you pass in an instance of a class with a `__call__` method defined. In this case `type(fn)` returns the class as a string.

@bveeramani @clarkzinzow @amogkam What do you prefer in this case? Should I use something like `f'map_batches: {type(fn)}'` as a fallback?

File: .bazeliskrc [hunk 5]
```diff
@@ -0,0 +1 @@
+USE_BAZEL_VERSION=5.x
```
Comment (bveeramani) [inline .bazeliskrc:1]: Remove this and `.gitignore` changes (?)

Also OOC did you run into Bazel issues?

File: python/ray/air/tests/test_dataset_config.py (hunk 3 above)
Comment (bveeramani) [reply python/ray/air/tests/test_dataset_config.py:253]: Is there any way we could use the `Preprocessor` name or representation? If you have many preprocessor I imagine it'd look something like this?

```
Stage 1->_transform_pandas->_transform_numpy->_transform_numpy->_transform_pandas
```

when you'd probably want something like this?

```
Stage 1->OneHotEncoder->Concatenator->TorchVisionPreprocessor->LabelEncoder
```

Comment (bveeramani) [reply python/ray/air/tests/test_dataset_config.py:253]: Actually, related to Eric's suggestion, maybe something like

```
MapBatches(my_udf) 
MapBatches(<lambda at 0x12346>)
```

for UDFs and 

```
OneHotEncoder  # Just use preprocessor name
Concatenator
```

for preprocessor?

I imagine it'd be more complex to implement, but I think the UX would be a lot better.

cc @ericl what do you think?

File: .bazeliskrc (hunk 5 above)
Comment (peytondmurray) [reply .bazeliskrc:1]: Whoops, sorry this snuck in. Yes, Bazel 6.0 fails to build this, see #31504. Swapping to bazelisk to control bazel works just fine.

File: python/ray/air/tests/test_dataset_config.py (hunk 1 above)
Comment (scottjlee) [reply python/ray/air/tests/test_dataset_config.py:288]: Yeah, I think either way works, I was just judging based off the existing test case. The existing test case checked `assert (string in stats)`, with the parentheses containing the expression that evaluates to a boolean value. In the case of `assert (string) in stats`, I think this also accomplishes the same thing. Feel free to follow your preference

Comment (peytondmurray) [reply python/ray/air/tests/test_dataset_config.py:288]: Wow, somehow I didn't spot this the first time around. Thanks for pointing this out, I think I'll change it back for consistency!

**Comment by peytondmurray** [comment]
@bveeramani @ericl Preprocessors now correctly show their class names; here's what the progress bars currently look like:

```
$ pytest -s test_batch_predictor.py::test_separate_gpu_stage
Test session starts (platform: linux, Python 3.10.8, pytest 7.2.0, pytest-sugar 0.9.6)
rootdir: /home/pdmurray/Desktop/workspace/ray, configfile: pytest.ini
plugins: clarity-1.0.1, sugar-0.9.6, typeguard-2.13.3, anyio-3.6.2
collecting ... 2023-01-23 09:14:16,764	INFO worker.py:1546 -- Started a local Ray instance. View the dashboard at 127.0.0.1:8265 
2023-01-23 09:14:18,290	WARNING read_api.py:330 -- ⚠️  The number of blocks in this dataset (10) limits its parallelism to 10 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
2023-01-23 09:14:18,298	INFO batch_predictor.py:184 -- `num_gpus_per_worker` is set for `BatchPreditor`.Automatically enabling GPU prediction for this predictor. To disable set `use_gpu` to `False` in `BatchPredictor.predict`.
Read->Dummypreprocessor: 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 12.83it/s]
2023-01-23 09:14:19,096	WARNING compute.py:543 -- `batch_size` is set to 4096, which reduces parallelism from 10 to 1. If the performance is worse than expected, this may indicate that the batch size is too large or the input block size is too small. To reduce batch size, consider decreasing `batch_size` or use the default in `map_batches`. To increase input block size, consider decreasing `parallelism` in read.
Map Progress (1 actors 1 pending): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.01s/it]
Shuffle Map: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 109.23it/s]
Shuffle Reduce: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 195.17it/s]
2023-01-23 09:14:20,129	WARNING read_api.py:330 -- ⚠️  The number of blocks in this dataset (10) limits its parallelism to 10 concurrent tasks. This is much less than the number of available CPU slots in the cluster. Use `.repartition(n)` to increase the number of dataset blocks.
Read: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 10/10 [00:00<00:00, 1568.43it/s]
2023-01-23 09:14:20,148	WARNING compute.py:543 -- `batch_size` is set to 4096, which reduces parallelism from 10 to 1. If the performance is worse than expected, this may indicate that the batch size is too large or the input block size is too small. To reduce batch size, consider decreasing `batch_size` or use the default in `map_batches`. To increase input block size, consider decreasing `parallelism` in read.
Map Progress (1 actors 1 pending): 100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:01<00:00,  1.98s/it]
Shuffle Map: 100%|█████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 155.07it/s]
Shuffle Reduce: 100%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████| 1/1 [00:00<00:00, 275.09it/s]

 python/ray/train/tests/test_batch_predictor.py ✓                                                                                                                                                  100% ██████████

Results (10.72s):
       1 passed
```

**Comment by peytondmurray** [comment]
Test failures appear unrelated.

File: python/ray/data/tests/test_dataset.py [hunk 6]
```diff
@@ -1517,26 +1517,28 @@ def test_dataset_repr(ray_start_regular_shared):
     assert repr(ds) == "Dataset(num_blocks=10, num_rows=10, schema=<class 'int'>)"
     ds = ds.map_batches(lambda x: x)
     assert repr(ds) == (
-        "MapBatches\n" "+- Dataset(num_blocks=10, num_rows=10, schema=<class 'int'>)"
+        "Mapbatches(<Lambda>)\n"
```
Comment (ericl) [inline python/ray/data/tests/test_dataset.py:1520]:
```suggestion
        "MapBatches(<Lambda>)\n"
```

We should use either snake or camel case here I think.

**Review by ericl — APPROVED** [review]
LGTM except the inconsistency of capitalization for mapbatches.

**Comment by ericl** [comment]
Oh title() is super weird, I don't know why anyone would want it force-lowercasing the other chars. How about we use this replacement function instead?

```
>>> def capfirst(s):
...     return s[:1].upper() + s[1:]
... 
>>> def capitalize(s):
...    return "".join(capfirst(x) for x in s.split("_"))
... 
>>> capitalize("map_batches")
'MapBatches'
>>> capitalize(capitalize("map_batches"))
'MapBatches'
```

**Comment by peytondmurray** [comment]
Sure, I'll try this and see what the impact on other tests is.

**Review by ericl — APPROVED** [review]

**Pull request merged**